File: sales/Mottivme Strategy 2026/_workspace/clientes/gohighlevel/repos/ghl-sdk-examples/php/sample-app-sql/storage/SqlStorage.php
<?php

namespace HighLevel\Storage;

use HighLevel\Storage\SessionStorage;
use HighLevel\Storage\SessionData;
use HighLevel\Logging\Logger;
use PDO;
use PDOException;
use Exception;

/**
 * SQL/MySQL implementation of SessionStorage
 * Stores OAuth tokens and session data in MySQL database
 * 
 * @package HighLevel\Storage
 */
class SqlStorage extends SessionStorage
{
    /**
     * PDO database connection
     * @var PDO|null
     */
    private ?PDO $client = null;

    /**
     * OAuth client ID
     * @var string|null
     */
    private ?string $clientId = null;

    /**
     * Database configuration
     * @var array
     */
    private array $config;

    /**
     * Default collection/table name
     * @var string
     */
    private string $defaultCollection = 'ghl_sessions';

    /**
     * Create new SqlStorage instance
     * 
     * @param string $host Database host
     * @param string $database Database name
     * @param string $username Database username
     * @param string $password Database password
     * @param string $port Database port
     * @param string $charset Database charset
     * @param Logger|null $logger Logger instance
     */
    public function __construct(
        string $host = 'localhost',
        string $database = 'ghl_sessions',
        string $username = 'root',
        string $password = '',
        string $port = '3306',
        string $charset = 'utf8mb4',
        ?Logger $logger = null
    ) {
        parent::__construct($logger);

        $this->config = [
            'host' => $host,
            'database' => $database,
            'username' => $username,
            'password' => $password,
            'port' => $port,
            'charset' => $charset
        ];
    }

    /**
     * Set the client ID
     * 
     * @param string $clientId OAuth client ID
     * @return void
     */
    public function setClientId(string $clientId): void
    {
        $this->clientId = $clientId;
    }

    /**
     * Initialize database connection and create tables
     * 
     * @return void
     * @throws Exception If connection fails
     */
    public function init(): void
    {
        try {
            $dsn = "mysql:host={$this->config['host']};port={$this->config['port']};dbname={$this->config['database']};charset={$this->config['charset']}";
            
            $this->client = new PDO($dsn, $this->config['username'], $this->config['password'], [
                PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
                PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
                PDO::ATTR_EMULATE_PREPARES => false,
            ]);

            $this->logger->debug('SqlStorage: Database connection established');
            
            // Create default sessions table
            $this->createCollection($this->defaultCollection);
            
        } catch (PDOException $e) {
            $this->logger->error('SqlStorage: Failed to connect to database: ' . $e->getMessage());
            throw new Exception('Failed to connect to database: ' . $e->getMessage());
        }
    }

    /**
     * Close database connection
     * 
     * @return void
     */
    public function disconnect(): void
    {
    $this->client = null;
        $this->logger->debug('SqlStorage: Database connection closed');
    }

    /**
     * Create sessions table if it doesn't exist
     * 
     * @param string $collectionName Table name
     * @return void
     * @throws Exception If table creation fails
     */
    public function createCollection(string $collectionName): void
    {
        try {
            $sql = "
                CREATE TABLE IF NOT EXISTS `{$collectionName}` (
                    `id` INT AUTO_INCREMENT PRIMARY KEY,
                    `resource_id` VARCHAR(255) NOT NULL UNIQUE,
                    `client_id` VARCHAR(255),
                    `access_token` TEXT,
                    `refresh_token` TEXT,
                    `token_type` VARCHAR(50),
                    `scope` TEXT,
                    `user_type` VARCHAR(50),
                    `company_id` VARCHAR(255),
                    `location_id` VARCHAR(255),
                    `expire_at` BIGINT,
                    `created_at` TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                    `updated_at` TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
                    INDEX `idx_resource_id` (`resource_id`),
                    INDEX `idx_client_id` (`client_id`),
                    INDEX `idx_expire_at` (`expire_at`)
                ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
            ";
            
            $this->client->exec($sql);
            $this->logger->debug("SqlStorage: Table '{$collectionName}' created or verified");
            
        } catch (PDOException $e) {
            $this->logger->error("SqlStorage: Failed to create table '{$collectionName}': " . $e->getMessage());
            throw new Exception("Failed to create table '{$collectionName}': " . $e->getMessage());
        }
    }

    /**
     * Get table reference (returns table name for SQL)
     * 
     * @param string $collectionName Table name
     * @return string Table name
     */
    public function getCollection(string $collectionName): string
    {
        return $collectionName;
    }

    /**
     * Store session data
     * 
     * @param string $resourceId Resource identifier
     * @param SessionData $sessionData Session data to store
     * @return void
     * @throws Exception If storage fails
     */
    public function setSession(string $resourceId, SessionData $sessionData): void
    {
        try {
            $sql = "
                INSERT INTO `{$this->defaultCollection}` (
                    resource_id, client_id, access_token, refresh_token, token_type,
                    scope, user_type, company_id, location_id, expire_at
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                ON DUPLICATE KEY UPDATE
                    access_token = VALUES(access_token),
                    refresh_token = VALUES(refresh_token),
                    token_type = VALUES(token_type),
                    scope = VALUES(scope),
                    user_type = VALUES(user_type),
                    company_id = VALUES(company_id),
                    location_id = VALUES(location_id),
                    expire_at = VALUES(expire_at),
                    updated_at = CURRENT_TIMESTAMP
            ";
            $expireAt = null;
            if (isset($sessionData->expires_in)) {
                $expireAt = $this->calculateExpireAt($sessionData->expires_in);
            }
            $stmt = $this->client->prepare($sql);
            $stmt->execute([
                $resourceId,
                $this->clientId,
                $sessionData->access_token,
                $sessionData->refresh_token,
                $sessionData->token_type,
                $sessionData->scope,
                $sessionData->user_type,
                $sessionData->company_id,
                $sessionData->location_id,
                $expireAt
            ]);

            $this->logger->debug("SqlStorage: Session stored for resource ID: {$resourceId}");
            
        } catch (PDOException $e) {
            $this->logger->error("SqlStorage: Failed to store session for {$resourceId}: " . $e->getMessage());
            throw new Exception("Failed to store session: " . $e->getMessage());
        }
    }

    /**
     * Retrieve session data
     * 
     * @param string $resourceId Resource identifier
     * @return SessionData|null Session data or null if not found
     */
    public function getSession(string $resourceId): ?SessionData
    {
        try {
            $sql = "SELECT * FROM `{$this->defaultCollection}` WHERE resource_id = ? AND (client_id = ? OR client_id IS NULL)";
            $stmt = $this->client->prepare($sql);
            $stmt->execute([$resourceId, $this->clientId]);
            $row = $stmt->fetch();

            if (!$row) {
                $this->logger->debug("SqlStorage: No session found for resource ID: {$resourceId}");
                return null;
            }

            // Create ISessionData from database row
            $sessionData = new SessionData();
            $sessionData->access_token = $row['access_token'];
            $sessionData->refresh_token = $row['refresh_token'];
            $sessionData->token_type = $row['token_type'];
            $sessionData->scope = $row['scope'];
            $sessionData->user_type = $row['user_type'];
            $sessionData->company_id = $row['company_id'];
            $sessionData->location_id = $row['location_id'];
            $sessionData->expire_at = $row['expire_at'];

            $this->logger->debug("SqlStorage: Session retrieved for resource ID: {$resourceId}");
            return $sessionData;
            
        } catch (PDOException $e) {
            $this->logger->error("SqlStorage: Failed to retrieve session for {$resourceId}: " . $e->getMessage());
            return null;
        }
    }

    /**
     * Delete session data
     * 
     * @param string $resourceId Resource identifier
     * @return void
     */
    public function deleteSession(string $resourceId): void
    {
        try {
            $sql = "DELETE FROM `{$this->defaultCollection}` WHERE resource_id = ? AND (client_id = ? OR client_id IS NULL)";
            $stmt = $this->client->prepare($sql);
            $stmt->execute([$resourceId, $this->clientId]);

            $this->logger->debug("SqlStorage: Session deleted for resource ID: {$resourceId}");
            
        } catch (PDOException $e) {
            $this->logger->error("SqlStorage: Failed to delete session for {$resourceId}: " . $e->getMessage());
        }
    }

    /**
     * Get access token only
     * 
     * @param string $resourceId Resource identifier
     * @return string|null Access token or null if not found
     */
    public function getAccessToken(string $resourceId): ?string
    {
        try {
            $sql = "SELECT access_token FROM `{$this->defaultCollection}` WHERE resource_id = ? AND (client_id = ? OR client_id IS NULL)";
            $stmt = $this->client->prepare($sql);
            $stmt->execute([$resourceId, $this->clientId]);
            $row = $stmt->fetch();

            return $row ? $row['access_token'] : null;
            
        } catch (PDOException $e) {
            $this->logger->error("SqlStorage: Failed to get access token for {$resourceId}: " . $e->getMessage());
            return null;
        }
    }

    /**
     * Get refresh token only
     * 
     * @param string $resourceId Resource identifier
     * @return string|null Refresh token or null if not found
     */
    public function getRefreshToken(string $resourceId): ?string
    {
        try {
            $sql = "SELECT refresh_token FROM `{$this->defaultCollection}` WHERE resource_id = ? AND (client_id = ? OR client_id IS NULL)";
            $stmt = $this->client->prepare($sql);
            $stmt->execute([$resourceId, $this->clientId]);
            $row = $stmt->fetch();

            return $row ? $row['refresh_token'] : null;
            
        } catch (PDOException $e) {
            $this->logger->error("SqlStorage: Failed to get refresh token for {$resourceId}: " . $e->getMessage());
            return null;
        }
    }

    /**
     * Get all sessions for current application
     * 
     * @return SessionData[] Array of session data
     */
    public function getSessionsByApplication(): array
    {
        try {
            $sql = "SELECT * FROM `{$this->defaultCollection}` WHERE client_id = ?";
            $stmt = $this->client->prepare($sql);
            $stmt->execute([$this->clientId]);
            $rows = $stmt->fetchAll();

            $sessions = [];
            foreach ($rows as $row) {
                $sessionData = new SessionData();
                $sessionData->access_token = $row['access_token'];
                $sessionData->refresh_token = $row['refresh_token'];
                $sessionData->token_type = $row['token_type'];
                $sessionData->scope = $row['scope'];
                $sessionData->user_type = $row['user_type'];
                $sessionData->company_id = $row['company_id'];
                $sessionData->location_id = $row['location_id'];
                $sessionData->expire_at = $row['expire_at'];
                
                $sessions[] = $sessionData;
            }

            return $sessions;
            
        } catch (PDOException $e) {
            $this->logger->error("SqlStorage: Failed to get sessions by application: " . $e->getMessage());
            return [];
        }
    }
}

File: sales/Mottivme Strategy 2026/_workspace/clientes/gohighlevel/projects/ghl-sdk-examples/python/sample-app-sql/storage/sql_storage.py
"""
SQL/MySQL implementation of SessionStorage for Python
Stores OAuth tokens and session data in MySQL database
"""

import pymysql
import time
from typing import Optional, Dict, Any, List
from highlevel.storage import SessionStorage, ISessionData
from highlevel.logging import Logger


class SqlStorage(SessionStorage):
    """
    SQL/MySQL implementation of SessionStorage
    Stores OAuth tokens and session data in MySQL database
    """

    def __init__(
        self,
        host: str = 'localhost',
        database: str = 'ghl_sessions',
        username: str = 'root',
        password: str = '',
        port: int = 3306,
        charset: str = 'utf8mb4',
        logger: Optional[Logger] = None
    ):
        """
        Create new SqlStorage instance

        Args:
            host: Database host
            database: Database name
            username: Database username
            password: Database password
            port: Database port
            charset: Database charset
            logger: Logger instance
        """
        super().__init__(logger)

        self.config = {
            'host': host,
            'database': database,
            'username': username,
            'password': password,
            'port': port,
            'charset': charset
        }

        self.client: Optional[pymysql.Connection] = None
        self.client_id: Optional[str] = None
        self.default_collection = 'ghl_sessions'

    def set_client_id(self, client_id: str) -> None:
        """Set the client ID"""
        self.client_id = client_id

    async def init(self) -> None:
        """
        Initialize database connection and create tables

        Raises:
            Exception: If connection fails
        """
        try:
            self.client = pymysql.connect(
                host=self.config['host'],
                port=self.config['port'],
                user=self.config['username'],
                password=self.config['password'],
                database=self.config['database'],
                charset=self.config['charset'],
                autocommit=True
            )

            self.logger.debug('SqlStorage: Database connection established')

            # Create default sessions table
            self.create_collection(self.default_collection)

        except pymysql.Error as e:
            self.logger.error(f'SqlStorage: Failed to connect to database: {e}')
            raise Exception(f'Failed to connect to database: {e}')

    async def disconnect(self) -> None:
        """Close database connection"""
        if self.client:
            self.client.close()
            self.client = None
        self.logger.debug('SqlStorage: Database connection closed')

    async def create_collection(self, collection_name: str) -> None:
        """
        Create sessions table if it doesn't exist

        Args:
            collection_name: Table name

        Raises:
            Exception: If table creation fails
        """
        try:
            with self.client.cursor() as cursor:
                sql = f"""
                    CREATE TABLE IF NOT EXISTS `{collection_name}` (
                        `id` INT AUTO_INCREMENT PRIMARY KEY,
                        `resource_id` VARCHAR(255) NOT NULL UNIQUE,
                        `client_id` VARCHAR(255),
                        `access_token` TEXT,
                        `refresh_token` TEXT,
                        `token_type` VARCHAR(50),
                        `scope` TEXT,
                        `user_type` VARCHAR(50),
                        `company_id` VARCHAR(255),
                        `location_id` VARCHAR(255),
                        `expire_at` BIGINT,
                        `created_at` TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                        `updated_at` TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
                        INDEX `idx_resource_id` (`resource_id`),
                        INDEX `idx_client_id` (`client_id`),
                        INDEX `idx_expire_at` (`expire_at`)
                    ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
                """

                cursor.execute(sql)
                self.logger.debug(f"SqlStorage: Table '{collection_name}' created or verified")

        except pymysql.Error as e:
            self.logger.error(f"SqlStorage: Failed to create table '{collection_name}': {e}")
            raise Exception(f"Failed to create table '{collection_name}': {e}")

    def get_collection(self, collection_name: str) -> str:
        """Get table reference (returns table name for SQL)"""
        return collection_name

    async def set_session(self, resource_id: str, session_data: ISessionData) -> None:
        """
        Store session data

        Args:
            resource_id: Resource identifier
            session_data: Session data to store

        Raises:
            Exception: If storage fails
        """
        try:
            expire_at = None
            if session_data.get('expires_in'):
                expire_at = self._calculate_expire_at(session_data.get('expires_in'))

            with self.client.cursor() as cursor:
                sql = f"""
                    INSERT INTO `{self.default_collection}` (
                        resource_id, client_id, access_token, refresh_token, token_type,
                        scope, user_type, company_id, location_id, expire_at
                    ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
                    ON DUPLICATE KEY UPDATE
                        access_token = VALUES(access_token),
                        refresh_token = VALUES(refresh_token),
                        token_type = VALUES(token_type),
                        scope = VALUES(scope),
                        user_type = VALUES(user_type),
                        company_id = VALUES(company_id),
                        location_id = VALUES(location_id),
                        expire_at = VALUES(expire_at),
                        updated_at = CURRENT_TIMESTAMP
                """

                cursor.execute(sql, [
                    resource_id,
                    self.client_id,
                    session_data.get('access_token'),
                    session_data.get('refresh_token'),
                    session_data.get('token_type'),
                    session_data.get('scope'),
                    session_data.get('userType'),
                    session_data.get('companyId'),
                    session_data.get('locationId'),
                    expire_at
                ])

            self.logger.debug(f"SqlStorage: Session stored for resource ID: {resource_id}")

        except pymysql.Error as e:
            self.logger.error(f"SqlStorage: Failed to store session for {resource_id}: {e}")
            raise Exception(f"Failed to store session: {e}")

    async def get_session(self, resource_id: str) -> Optional[ISessionData]:
        """
        Retrieve session data

        Args:
            resource_id: Resource identifier

        Returns:
            Session data or None if not found
        """
        try:
            with self.client.cursor(pymysql.cursors.DictCursor) as cursor:
                sql = f"SELECT * FROM `{self.default_collection}` WHERE resource_id = %s AND (client_id = %s OR client_id IS NULL)"
                cursor.execute(sql, [resource_id, self.client_id])
                row = cursor.fetchone()

                if not row:
                    self.logger.debug(f"SqlStorage: No session found for resource ID: {resource_id}")
                    return None

                session_data = ISessionData(
                    access_token=row.get('access_token'),
                    refresh_token=row.get('refresh_token'),
                    token_type=row.get('token_type'),
                    scope=row.get('scope'),
                    userType=row.get('user_type'),
                    companyId=row.get('company_id'),
                    locationId=row.get('location_id')
                )
                session_data.expire_at = row.get('expire_at')

                self.logger.debug(f"SqlStorage: Session retrieved for resource ID: {resource_id}")
                return session_data

        except pymysql.Error as e:
            self.logger.error(f"SqlStorage: Failed to retrieve session for {resource_id}: {e}")
            return None

    async def delete_session(self, resource_id: str) -> None:
        """
        Delete session data

        Args:
            resource_id: Resource identifier
        """
        try:
            with self.client.cursor() as cursor:
                sql = f"DELETE FROM `{self.default_collection}` WHERE resource_id = %s AND (client_id = %s OR client_id IS NULL)"
                cursor.execute(sql, [resource_id, self.client_id])

            self.logger.debug(f"SqlStorage: Session deleted for resource ID: {resource_id}")

        except pymysql.Error as e:
            self.logger.error(f"SqlStorage: Failed to delete session for {resource_id}: {e}")

    async def get_access_token(self, resource_id: str) -> Optional[str]:
        """
        Get access token only

        Args:
            resource_id: Resource identifier

        Returns:
            Access token or None if not found
        """
        try:
            with self.client.cursor() as cursor:
                sql = f"SELECT access_token FROM `{self.default_collection}` WHERE resource_id = %s AND (client_id = %s OR client_id IS NULL)"
                cursor.execute(sql, [resource_id, self.client_id])
                row = cursor.fetchone()

                return row[0] if row else None

        except pymysql.Error as e:
            self.logger.error(f"SqlStorage: Failed to get access token for {resource_id}: {e}")
            return None

    async def get_refresh_token(self, resource_id: str) -> Optional[str]:
        """
        Get refresh token only

        Args:
            resource_id: Resource identifier

        Returns:
            Refresh token or None if not found
        """
        try:
            with self.client.cursor() as cursor:
                sql = f"SELECT refresh_token FROM `{self.default_collection}` WHERE resource_id = %s AND (client_id = %s OR client_id IS NULL)"
                cursor.execute(sql, [resource_id, self.client_id])
                row = cursor.fetchone()

                return row[0] if row else None

        except pymysql.Error as e:
            self.logger.error(f"SqlStorage: Failed to get refresh token for {resource_id}: {e}")
            return None

    async def get_sessions_by_application(self) -> List[ISessionData]:
        """
        Get all sessions for current application

        Returns:
            List of session data
        """
        try:
            with self.client.cursor(pymysql.cursors.DictCursor) as cursor:
                sql = f"SELECT * FROM `{self.default_collection}` WHERE client_id = %s"
                cursor.execute(sql, [self.client_id])
                rows = cursor.fetchall()

                sessions = []
                for row in rows:
                    session_data = ISessionData(
                        access_token=row.get('access_token'),
                        refresh_token=row.get('refresh_token'),
                        token_type=row.get('token_type'),
                        scope=row.get('scope'),
                        user_type=row.get('user_type'),
                        company_id=row.get('company_id'),
                        location_id=row.get('location_id')
                    )
                    session_data.expire_at = row.get('expire_at')
                    sessions.append(session_data)

                return sessions

        except pymysql.Error as e:
            self.logger.error(f"SqlStorage: Failed to get sessions by application: {e}")
            return []

    def _calculate_expire_at(self, expires_in: int) -> int:
        """Calculate expiration timestamp"""
        return int(time.time() * 1000) + (expires_in * 1000)
